File: k-kbk/6주차/14889.py
from itertools import combinations

n = int(input())

nums = [list(map(int, input().split(" "))) for _ in range(n)]
members = list(range(n))
result = 100 * 20

for startTeam in combinations(members, n // 2):
    start, link = 0, 0
    linkTeam = list(set(members) - set(startTeam))

    for c in combinations(startTeam, 2):
        start += nums[c[0]][c[1]]
        start += nums[c[1]][c[0]]

    for c in combinations(linkTeam, 2):
        link += nums[c[0]][c[1]]
        link += nums[c[1]][c[0]]

    result = min(result, abs(start - link))

print(result)

File: keemsebin/6주차/14889.js
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let input = [];

rl.on("line", function (line) {
  input.push(line);
}).on("close", function () {
  let N = Number(input[0]);
  let arr = input.slice(1).map((item) => item.split(" ").map(Number));
  const check = new Array(N).fill(false);

  let teamNum = N / 2;
  let minDiff = Infinity;

  const calculateTeamScore = (team) => {
    let score = 0;
    for (let i = 0; i < team.length; i++) {
      for (let j = i + 1; j < team.length; j++) {
        score += arr[team[i]][team[j]] + arr[team[j]][team[i]];
      }
    }
    return score;
  };

  const dfs = (idx, count) => {
    if (count === teamNum) {
      let startTeam = [];
      let linkTeam = [];

      for (let i = 0; i < N; i++) {
        if (check[i]) {
          startTeam.push(i);
        } else {
          linkTeam.push(i);
        }
      }

      let startScore = calculateTeamScore(startTeam);
      let linkScore = calculateTeamScore(linkTeam);
      let diff = Math.abs(startScore - linkScore);
      minDiff = Math.min(minDiff, diff);
      return;
    }

    for (let i = idx; i < N; i++) {
      if (!check[i]) {
        check[i] = true;
        dfs(i + 1, count + 1);
        check[i] = false;
      }
    }
  };

  dfs(0, 0);
  console.log(minDiff);
});
